lib/network_manager.go

```diff
@@ -0,0 +1,787 @@
+// the handshake process, or became invalid for some other reason.
+type NetworkManager struct {
+	// The parameters we are initialized with.
+	params *DeSoParams
+
+	cmgr        *ConnectionManager
+	blsKeystore *BLSKeystore
+
+	handshake *HandshakeManager
+	rnManager *RemoteNodeManager
+
+	// The address manager keeps track of peer addresses we're aware of. When
+	// we need to connect to a new outbound peer, it chooses one of the addresses
+	// it's aware of at random and provides it to us.
+	AddrMgr *addrmgr.AddrManager
+
+	// When --connect-ips is set, we don't connect to anything from the addrmgr.
+	connectIps []string
+	// persistentIpToRemoteNodeIdsMap maps persistent IP addresses, like the --connect-ips, to the RemoteNodeIds of the
+	// corresponding RemoteNodes. This is used to ensure that we don't connect to the same persistent IP address twice.
+	// And that we can reconnect to the same persistent IP address if we disconnect from it.
+	persistentIpToRemoteNodeIdsMap *collections.ConcurrentMap[string, RemoteNodeId]
+
+	activeValidatorsMapLock sync.RWMutex
+	// activeValidatorsMap is a map of all currently active validators registered in consensus. It will be updated
+	// periodically by the owner of the NetworkManager.
+	activeValidatorsMap *collections.ConcurrentMap[bls.SerializedPublicKey, consensus.Validator]
+
+	// The target number of non-validator outbound remote nodes we want to have. We will disconnect remote nodes once
+	// we've exceeded this number of outbound connections.
+	targetNonValidatorOutboundRemoteNodes uint32
+	// The target number of non-validator inbound remote nodes we want to have. We will disconnect remote nodes once
+	// we've exceeded this number of inbound connections.
+	targetNonValidatorInboundRemoteNodes uint32
+	// When true, only one connection per IP is allowed. Prevents eclipse attacks
+	// among other things.
+	limitOneInboundRemoteNodePerIP bool
+
+	startGroup sync.WaitGroup
+	exitChan   chan struct{}
+	exitGroup  sync.WaitGroup
+}
+
+func NewNetworkManager(params *DeSoParams, cmgr *ConnectionManager, rnManager *RemoteNodeManager,
+	blsKeystore *BLSKeystore, addrMgr *addrmgr.AddrManager, connectIps []string,
+	targetNonValidatorOutboundRemoteNodes uint32, targetNonValidatorInboundRemoteNodes uint32,
+	limitOneInboundConnectionPerIP bool) *NetworkManager {
+
+	return &NetworkManager{
+		params:                                params,
+		cmgr:                                  cmgr,
+		blsKeystore:                           blsKeystore,
+		handshake:                             NewHandshakeController(rnManager),
+		rnManager:                             rnManager,
+		AddrMgr:                               addrMgr,
+		connectIps:                            connectIps,
+		persistentIpToRemoteNodeIdsMap:        collections.NewConcurrentMap[string, RemoteNodeId](),
+		activeValidatorsMap:                   collections.NewConcurrentMap[bls.SerializedPublicKey, consensus.Validator](),
```

Comment:
So where would this be set initially, and where would it be updated? It feels like we want to set it in Server but only after we've gotten our tip up-to-date. Then in the steady-state it can be set after every epoch update.

Reply:
@diamondhands0 - the activeValidatorsMap is set in FastHotStuffConsensus when transitioning to the steady state and when updating the tip block. Both are cheap and result in minimal additional overhead since we have to fetch the validator set to verify every block